File: cjaewon/week02/BOJ_20922_겹치는 건 싫어.py
n, k = map(int, input().split()) # n: 수열의 개수 k: 중복 허용 수
arr = list(map(int, input().split()))

count = {}

# n이 20만이므로 시간 초과가 나지 않도록 주의해야함
# 어떤 방식으로 접근해야할까
# 무지성으로 품...
# 어떤 숫자가 특정 개수를 초과할 시 그 숫자가 처음 나오는 곳으로 이동한 후 한 칸 이동해서 다시 갯수 세기 반복

i = 0
length = 0
max_length = 0

while i < n:
  if arr[i] in count:
    count[arr[i]][0] += 1
    if count[arr[i]][0] > k:
      i = count[arr[i]][1] + 1
      count = {}
      length = 0
      continue
  else:
    count[arr[i]] = [1, i] # 수의 중복 개수, 수의 가장 최근 index

  length += 1
  i += 1

  if length > max_length:
    max_length = length

print(max_length)

File: Leeseojun035/week02/BOJ_20922_겹치는 건 싫어.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    int N, K;
    cin >> N >> K;
    
    vector<int> arr(N);
    vector<int> count(100001, 0);  // 숫자 등장 횟수 저장 
    
    for (int i = 0; i < N; i++) cin >> arr[i];

    int left = 0, right = 0;
    int max_length = 0;

    while (right < N) {
        count[arr[right]]++;

        // 만약 현재 숫자의 개수가 K를 초과하면 left를 이동시켜 제한을 맞춤
        while (count[arr[right]] > K) {
            count[arr[left]]--;
            left++;
        }

        // 현재 연속 부분 수열의 길이 갱신
        max_length = max(max_length, right - left + 1);

        right++;  // 다음 숫자로 이동
    }

    cout << max_length << '\n';
    return 0;
}
